Implementation to calculate conditional independencies

cthoyt opened the pull request:
Closes #24

This pull request implements an algorithm to calculate all conditional independencies in a given ADMG.

- [x] Add outline of implementation, signature, and tests (@cthoyt; 1140eca)
- [x] Add additional unit tests (@y0-causal-inference/y0-team, following #26)
- [x] Implementation (@djinnome or @JosephCottam, please assign yourself)
- [x] Remove references to the dsl in ConditionalIndependency
- [x] Merge DSeparationJudgement and ConditionalIndependency

cthoyt commented on tests/test_algorithm/test_get_conditional_independencies.py:
```diff
@@ -0,0 +1,31 @@
+from y0.graph import NxMixedGraph, napkin_graph
+
+
+class TestGetConditionalIndependencies(unittest.TestCase):
```
other tests welcome! Likely better to follow #26 when more graphs are available and this PR won't blow up

cthoyt replied:
Done. now tests can be arbitrarily added as instances of the `Example` class in `y0.examples`

cthoyt commented on src/y0/algorithm/falsification.py:
```diff
@@ -0,0 +1,114 @@
+from ..util import CITests
```
can we make `pgmpy` a requirement and just import the functions in CITests directly from there?

JosephCottam replied:
I wanted to be conservative in the number of dependencies.  We really only need this one file (and it could be reduced down to one function from this file) at this time.

cthoyt replied:
Excellent, then we can just copy it (just double check the licensing is okay and we make a very specific note about it)

JosephCottam replied:
Its under MIT license, which I copied to the top of the file and have note that says where it was originally copied from.

cthoyt replied:
Perfect. Thank you!

cthoyt commented on tests/test_algorithm/test_get_conditional_independencies.py:
```diff
@@ -0,0 +1,63 @@
+    "Test the d-separation utility."
+
+    # TODO: Migrate to the ADMG representation, not just vanilla networkx
+    def test_mit_example(self):
```
it would be great to curate examples from the example graphs in `y0.examples`

JosephCottam replied:
I can do that.

JosephCottam commented on src/y0/algorithm/falsification.py:
```diff
@@ -90,24 +90,16 @@ def __init__(self, failures, evidence):
     all_pairs = combinations(all_nodes, 2)
 
-    to_test = [(a, b, given) for a,b in wrapper(all_pairs, desc="Checking d-separation")
+    to_test = [(a, b, given) for a,b in tqdm(all_pairs, desc="Checking d-separation")
```
I would like to leave the verbose flag in place, only using tqdm when verbose is requested.  
Moving tqdm to the the top is a good idea, since it is a project dependency elsewhere.

cthoyt replied:
tqdm has a keyword argument in it you can use instead to disable it called `disable`

JosephCottam replied:
Was not aware of that.  Great!

cthoyt commented on src/y0/examples.py:
```diff
@@ -514,4 +515,19 @@ class Example:
+d_separation_example = Example(
+    name="D-separation example",
+    reference="http://web.mit.edu/jmn/www/6.034/d-separation.pdf",
+    graph=NxMixedGraph.from_edges(
```
maybe we can make a field in the Example for example d-separations to test

JosephCottam replied:
I don't think that's necessary: The expected conditional independencies are the things that are d-separated.  If we added the field, it would just cary duplicate information.  

In the long run, we might rename the field and related class, but for now I think the existing framework works.

cthoyt replied:
oh, thank you for the insight! Sorry that it's up to you guys to fill in the gaps in my knowledge 😁

cthoyt commented on src/y0/algorithm/falsification.py:
```diff
@@ -0,0 +1,153 @@
+    return rv
+
+
+def all_combinations(source, min: int = 0, max: Optional[int] = None):
```
need unit tests for this function

cthoyt commented on src/y0/algorithm/falsification.py:
```diff
@@ -0,0 +1,156 @@
+    return rv
+
+
+def get_augments(graph: SG):
```
unit tests

cthoyt commented on src/y0/algorithm/falsification.py:
```diff
@@ -0,0 +1,156 @@
+
+def get_augments(graph: SG):
+    parents = [graph.parents([v]) for v in graph.vertices]
+    augments = [*chain(*[combinations(nodes, 2) for nodes in parents if len(parents) > 1])]
```
In general rather than doing `[*expr]` you can just use the `list(expr)` function. Same can be said that `chain(*expr)` is the same as `chain.from_iterable(expr)`

```suggestion
    auguments = list(chain.from_iterable(combinations(nodes, 2) for nodes in parents if len(parents) > 1]))
```

JosephCottam replied:
I generally prefer the `[*...]` notation to the `list(...)` notation.  Is there reason why you would discourage the symbolic notation?

`chain.from_iterable` makes sense.

cthoyt replied:
Not all readers will understand `[*...]` and it's uncommonly used, which further makes reading the code harder to understand as a whole

cthoyt replied:
I hope you don't mind I was doing a bit of maintenance and rewriting parts as I was trying to figure them out

cthoyt commented:
@JosephCottam what's left on the to-do list for this PR? Maybe you could edit the description to leave the remaining tasks. I'd rather get some working and tested functionality merged then continue work in an additional PR. @djinnome any thoughts?

cthoyt commented:
@JosephCottam I don't know if we need to merge the DSeparationJudgement and ConditionalIndependency - I think it's enough that their shared functionality is reused. Rather than merging, perhaps the DSeparationJudgement can contain an instance of a ConditionalIndependency since there is more information necessary for a DSeparationJudgement in addition to the ConditionalIndependency

JosephCottam commented:
@cthoyt Increasing the consistency and interplay between ConditionalIndependency and the DSeparationJudgement classes makes sense.  We can always revisit later.  I'll take a stab at it today!

cthoyt commented on src/y0/algorithm/conditional_independencies.py:
```diff
@@ -0,0 +1,173 @@
+        policy = _len_lex if policy is None else policy
+
+        dseps = sorted(dseps, key=_grouper)
+        instances = {k: sorted(vs, key=policy)[0]
```
`sorted(x, key=y)[0]` should be the same as `min(x, key=y)`

JosephCottam replied:
Sure. Changed.

cthoyt commented on tests/test_algorithm/test_get_conditional_independencies.py:
```diff
@@ -0,0 +1,86 @@
+                            f"Conditional independency is not a d-separation in {graph}")
+
+        gist = {(ci.left, ci.right) for ci in subject}
+        self.assertEqual(len(gist), len(set(gist)), "Duplicate left/right pair observed")
```
since gist is a set this will always be equal. maybe define gist as a list comprehension instead?

JosephCottam replied:
Ha, yes. Copy-paste error.  Fixed.

JosephCottam commented:
I ended up merging the d-separation (DS) and conditional independency (CI) classes.  The reason is that CI kept feeling like a DS in a specific context. More specifically a CI is a DS that belongs to a graphs minimal set of DSs given a policy.  At least, that's how it's being used right now.  We can split them apart later if we can make a more clean semantic statement about what each is.  As is, tests that passed before pass now.  

@cthoyt Can you repair the test-case that refers to invalid nodes in a CI?  The reference isn't sufficient for me to know where the graph came from so I can't do it myself; I don't know if it's the graph definition or the CI definition that is in error.

cthoyt commented on .flake8:
```diff
@@ -29,4 +29,4 @@ import-order-style = pycharm
 application-import-names =
     y0
     tests
-format = ${cyan}%(path)s${reset}:${yellow_bold}%(row)d${reset}:${green_bold}%(col)d${reset}: ${red_bold}%(code)s${reset} %(text)s
```
is there a reason for this change?

JosephCottam replied:
In the terminal, the character number was much more visible than the line-number.  This change made the line-number more emphasized. 

These days I'm actually using `format = ${red}%(code)s${reset} ${cyan}%(path)s${reset}:${red_bold}%(row)d${reset} %(text)s`.  Which doesn't give the character number at all and moves the error-code to the left (where it is easier to ignore because its rarely needed).

cthoyt commented on src/y0/algorithm/conditional_independencies.py:
```diff
@@ -0,0 +1,144 @@
+
+
+def get_conditional_independencies(
+    graph: SG,
```
the reason we're using this style of definitions throughout the package and not the hanging one is that it's consistent in where the variables line up regardless of the length of the function name. the 4 space indent also is consistent with all of the other indentation in python code, while hanging indents make it much more difficult to read - that's why I've gone back and changed these again

cthoyt commented:
I found some typos in the identifiability 2 example and fixed those, but errors still remain. I changed the tests to look at the judgements themselves, and now it's obvious that there's quite a big discrepancy between the results coming out and what CausalFusion says. Here is the example for Identifiability 2 (sorry there's no way to link to this page in the CausalFusion app)

<img width="554" alt="Screen Shot 2021-03-06 at 16 22 07" src="https://user-images.githubusercontent.com/5069736/110211700-1a853b80-7e98-11eb-8a06-2b2496d4d54d.png">

It's also the case that when making the test more general that the identifiability 1 graph also begins to fail. Have I made a mistake in that these aren't actually supposed to correspond to each other?

JosephCottam commented:
My current understanding is that each canonical left/right pair of identified D-separations, however when there are multiple d-separation conditions it is a choice which on you use.  Therefore, our output doesn't need to strictly duplicate what is in causalfusion to be correct, but there is a correspondence.  

For example the graph A->B->C->D has d-separations (among others):
* A _||_ D | B
* A _||_ D | C
* A _||_ D | B,C

Our implementation needs to pick one for conditional independency for A _||_ B and none inherently wrong. Causal Fusion may choose a different one than we do, as long as we choose one. Our default-selection policy would choose the first one (shortest and lexicographic first).  Working with causal fusion, it sometimes changes its choice based on labels or unrelated topological changes (like add a new node to a long un-related chain) and I don't see a benefit to reverse-engineering  their algorithm's policy.   

With a corrected topology, I was going to go through and verify that (1) every conditional independency left/right pair that causal fusion came up with existed in our set and (2) that the condition that our algorithm picked are in fact d-separations.

cthoyt commented:
> Our implementation needs to pick one for conditional independency for A _||_ B and none inherently wrong. Causal Fusion may choose a different one than we do, as long as we choose one. Our default-selection policy would choose the first one (shortest and lexicographic first). Working with causal fusion, it sometimes changes its choice based on labels or unrelated topological changes (like add a new node to a long un-related chain) and I don't see a benefit to reverse-engineering their algorithm's policy.

If that's the case, let's split the function that's doing the work in two. The first bit should return all possible d-separations, and the second one should take that result (and whatever else it needs) and decide on which ones it wants to report.

Thanks for letting me know about this discrepancy. If that's the case, then the previous less strict testing made more sense. But, how would we test that the givens are valid without explicitly enumerating them all, before a policy is applied for choosing?

JosephCottam commented:
Our code does do it in two stages already! `d_separations` produces all d-separations, and `get_condition_independencies` filters it to the appropriate set.  

In truth, the the `minimal` function implements the guts of filtering and `get_conditional_independency` is a wrapper that threads the two parts together and does graph-conversion when required.  The `policy` argument to `minimal` provides the flexibility for a specific policy to use. I did not include it in `get_conditional_independency` but it could be threaded through trivially.

For testing, I think enumerating all d-separations for anything but trivial graphs is going to be annoying.  There are probably other ways to test it that will at least alert us if we either (1) change the policy or (2) break  d-separation or (3) break the minimizing filter.  This will give us something of a "canary" (warning when something is wrong) so errors won't propagate, but won't always tell us exactly what is wrong.  With the correct test graph, I can take a pass at it later today.

cthoyt commented:
@JosephCottam if you're sure about the results, then please go ahead and correct the tests yourself. Perhaps we could write some more specific tests that show off how different policies could work (maybe later, if we have a reason to use a non-default policy). So in the end, will results ever be different based on which CIs are chosen?

JosephCottam commented:
It is possible that results would be different based on data quality or
causal graph correctness.

For example:
Given a model A->B->C->D->E, but a true causal graph with an edge from C to
E as well.
If you check A _||_ E | B, the conditional independence would hold BUT if
you condition A_||_E | D it probably will not.  For this reason, a good
default policy is probably topologically highest, since it would give the
most discriminatory information (since the error would show up when it also
checked C _||_ E | D) .  Topology-based requires an extra parameter (the
graph) to the minimize function.  I have a draft of that locally, but it's
not right yet so I haven't checked it in!



On Mon, Mar 8, 2021 at 2:22 PM Charles Tapley Hoyt <notifications@github.com>
wrote:

> @JosephCottam <https://github.com/JosephCottam> if you're sure about the
> results, then please go ahead and correct the tests yourself. Perhaps we
> could write some more specific tests that show off how different policies
> could work (maybe later, if we have a reason to use a non-default policy).
> So in the end, will results ever be different based on which CIs are chosen?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/y0-causal-inference/y0/pull/29#issuecomment-793125349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWTUFI4GCOV74XTFMJSYXLTCVEZPANCNFSM4WNZV2MA>
> .
>

JosephCottam commented:
@cthoyt I have addressed most of the formatting issues. 

@cthoyt  @djinnome I move that we ignore error N400 (backslash for line continuation).   The  PEP8 description centers on expressions with operator chaining.  However, it is used for function chaining here.  Though the warning can be cleared by adding parenthesis around the whole chain (see below), I think the line continuation version is more clear in what is going on.

```
    evidence = (pd.DataFrame(rows, columns=["left", "right", "given", "chi^2", "p", "dof"])
        .sort_values("p")
        .assign(**{"Holm–Bonferroni level": significance_level / pd.Series(range(len(rows) + 1, 0, -1))})
        .pipe(lambda df: df.assign(flagged=(df["p"] < df["Holm–Bonferroni level"])))
        .sort_values(["flagged", "dof"], ascending=False))```

cthoyt commented on notebooks/Falsification.ipynb:
```diff
@@ -0,0 +1,887 @@
+{
```
@JosephCottam is this notebook just for exploration? I think, if so it should be removed.

cthoyt replied:
If you want to write more thorough user documentation, the best place to put it is in the module-level docstring or in the `docs/source/` folder

cthoyt replied:
I think it would be better to write these up as documentation instead of notebooks - the problem is jupyter notebooks can get out of date very quickly and are not readily able to be combined with other documentation

JosephCottam replied:
Falsification is actually a down-stream of conditional independency testing (issue #23).  I don't think its ready for a fully-fledged example yet. If you prefer not to use a jupyter notebook example, then I'm fine with it being removed.

cthoyt replied:
Let's just remove it for now, then

cthoyt commented:
@JosephCottam it looks like some of the tests started failing, but not related to code I changed. Are these non-deterministic, or is there something I missed?

JosephCottam commented:
@cthoyt To the best of my knowledge, all of the tests are deterministic.

JosephCottam commented:
@cthoyt Fixed the test case that started failing.  The changes made to loading the asia example data broke that test case.  It was just putting the _file path_ into the dataframe, rather than loading the data from the file.  Now it loads the data again!

cthoyt commented:
Thank you so much @JosephCottam for all of the effort! Looking forward to what's next
